PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+						current = current.node_ref.ascend().parent;
+						delete to_delete.node->as_branch();
+					}
+				}
+			}
+
+			LeafNode* new_root_to_be = current.node_ref.node;
+
+			auto ascension = current.node_ref.ascend();
+			while (ascension.succeeded) {
+				KVERef node = ascension.parent;
+				ascension = node.node_ref.ascend();
+				delete node.node_ref.node->as_branch();
+			}
+
+			//reinitialize the root node
+			new_root_to_be->~LeafNode();
+			::new(new_root_to_be) LeafNode();
+
+			this->root = {new_root_to_be, 0};
+		}
+
+	public:
+		Entry Entry(const K& key) {
+			const auto search_result = search_tree(this->root.as_node_ref(), key);
+			switch (search_result.result) {
+				case SearchResult::FOUND:  return Entry::Occupied(search_result.handle, this->len);
+				case SearchResult::FAILED: return Entry::Vacant(  search_result.handle, this->len, key);
+				default: UNREACHABLE();
+			}
+		}
+		Range<K, V, Bfactor> MaxRange() { return {BTree::first_leaf_edge(this->root.as_node_ref()), BTree::last_leaf_edge(this->root.as_node_ref())}; }
+		Range<K, V, Bfactor> GetRange(const K& lower, const K& upper) { ASSERTE(lower < upper, ""); return {search_tree(this->root.as_node_ref(), lower).handle, search_tree(this->root.as_node_ref(), upper).handle}; }
+		Iter begin() { return {BTree::first_leaf_edge(this->root.as_node_ref()), this->GetSize(), this->GetSize()}; }
+		Iter end()   { return {BTree:: last_leaf_edge(this->root.as_node_ref()), 0,               this->GetSize()}; }
+		class Keys   Keys()   { return {*this}; }
+		class Values Values() { return {*this}; }
+
+	public:
+		void swap(BTreeMap& other) { std::swap(this->root, other.root); std::swap(this->len, other.len); }
+
+	private:
+		template<typename Key, typename Val>
+		Optional<V> insert(Key&& key, Val&& value) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (entry.IsVacant()) {
+				entry.VacantInsert(std::forward<Val>(value));
+				return {};
+			} else {
+				return {entry.OccupiedInsert(std::forward<Val>(value))};
+			}
+		}
+
+		template<typename Key>
+		Optional<V> remove(Key&& key) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (!entry.IsVacant()) {
+				return {entry.Remove()};
+			} else {
+				return {};
+			}
+		}
+
+		template<typename Key>
+		Optional<V&> get(Key&& key) {
+			auto entry = this->Entry(std::forward<Key>(key));
+			if (!entry.IsVacant()) {
+				return {entry.OccupiedGet()};
+			} else {
+				return {};
+			}
+		}
+
+	private:
+		class Entry	{
+		public:
+			static Entry Vacant  (KVERef kv_ref, size_t& len, const K&  key) { return Entry(kv_ref, len, key); }
+			static Entry Vacant  (KVERef kv_ref, size_t& len,       K&& key) { return Entry(kv_ref, len, std::move(key)); }
+			static Entry Occupied(KVERef kv_ref, size_t& len)                { return Entry(kv_ref, len); }
+
+		public:
+			bool IsVacant() { return static_cast<bool>(key); }
+
+			V& OrInsert(const V&  value) { return this->or_insert(value); }
```

Comment:
Methods called without `this->` confuse the hell out of me as I think them to be non-member functions at first :( I think it's better to be explicit, but I will remove these if you insist.